### PR TITLE
Rnacentral updates

### PR DESCRIPTION
**Pull request by tgurbich**
- Brought the RNAcentral script up to date with the pipeline changes (including the switch to tRNAscan-SE)
- Added accounting for private samples
- Added accounting for previous catalogue versions
- Added unit tests

### CODE REVIEW

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py [hunk 1]
```diff
@@ -259,38 +325,48 @@ def get_publications(genome_sample_accession, reported_project, insdc_accession)
     else:
         # keep going through levels of samples until we get to the raw read samples
         while samples_to_check:
+            infinite_loop_protection += 1
+            if infinite_loop_protection > 20:
```
Comment (mberacochea) [inline helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:329]: Maybe 20 is too much? If loads of entries get to more than 10 it will slow down the process a lot, right?

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py [hunk 2]
```diff
@@ -126,9 +180,12 @@ def generate_data_dict(mgnify_accession, sample_accession, taxonomy, deoverlap_d
     global SKIP_CMSCAN, SKIP_GFF, GOOD, BAD_SEQUENCE
     deoverlap_path = os.path.join(deoverlap_dir, "{}.cmscan-deoverlap.tbl".format(mgnify_accession))
     if not os.path.exists(deoverlap_path):
-        logging.warning("cmscan file for accession {} doesn't exist. Skipping.".format(mgnify_accession))
-        SKIP_CMSCAN += 1
-        return None
+        # check if the naming has changed
+        deoverlap_path = os.path.join(deoverlap_dir, "{}.ncrna.deoverlap.tbl".format(mgnify_accession))
+        if not os.path.exists(deoverlap_path):
+            logging.warning("cmscan file for accession {} doesn't exist. Skipping.".format(mgnify_accession))
```
Comment (mberacochea) [inline helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:186]: Please add the expected path to the log. It is easier to debug if you have the expected path available, especially when something is not supposed to happen.

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py [hunk 3]
```diff
@@ -259,38 +325,48 @@ def get_publications(genome_sample_accession, reported_project, insdc_accession)
...
+                                "SAMPLE_ATTRIBUTE"]
+                        else:
+                            # there is no XML for this sample, we can't get samples it is derived from
+                            sample_attributes = list()
+                    except:
+                        # There is XML but its format is wrong and we can't parse it
+                        logging.exception("Unable to process XML for sample {}".format(sample_to_check))
+                        sys.exit()
+                    sample_is_derived = False
+                    for attribute in sample_attributes:
```
Comment (mberacochea) [inline helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:352]: For a future version, I think we should considering storing all of the metadata in a sqlite db instead of flat files.

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py [hunk 4]
```diff
@@ -428,16 +507,30 @@ def get_project_accession(biosample):
...
+        return None
     xml_url = 'https://www.ebi.ac.uk/ena/browser/api/xml/{}'.format(sample_id)
     try:
         r = run_xml_request(xml_url)
     except:
-        if insdc_accession.startswith("C"):
-            logging.error("Unable to get XML for sample {}, ENA genome {}".format(sample_id, insdc_accession))
-            sys.exit()
+        if r is None:
+            if insdc_accession.startswith("C"):
```
Comment (mberacochea) [inline helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:519]: Why "C"?

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py [hunk 5]
```diff
@@ -428,16 +507,30 @@ def get_project_accession(biosample):
...
+                logging.error("Unable to get XML for sample {}, ENA genome {}".format(sample_id, insdc_accession))
+                sys.exit()
```
Comment (mberacochea) [inline helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:521]: Should the script exit here? Should just print the error and skip the genome?

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py [hunk 6]
```diff
@@ -449,14 +542,42 @@ def load_xml(sample_id, insdc_accession=None):
...
+                ERROR_404.append(sample_id)
+        else:
+            logging.error("Error in full: {}".format(r.text))
+            sys.exit("Program is exiting - unable to query ENA for sample {}. Error code {}".format(sample_id, 
+                                                                                                    r.status_code))
         return None
 
 
-@retry(tries=5, delay=10, backoff=1.5)
 def run_xml_request(xml_url):
```
Comment (mberacochea) [inline helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:556]: Why did you remove retry? It wasn't working properly, or just to remove the dependency?

**Review by mberacochea — APPROVED** [review]
Thanks, @tgurbich. I left a few questions/comments, but I think it's good to merge, in my opinion.

We plan to integrate this into Nextflow, and at that point, we should do some refactoring (such as keeping track of things with SQLite instead of flat files, which will be easier to maintain and audit in the future). Overall, the code is gud; I'm not a fan of the GLOBALS, but in this synchronous code, it should be alright.

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py (hunk 1 above)
Comment (tgurbich) [reply helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:329]: This is only meant to break an (unlikely) infinite loop. I can reduce it, chose 20 just to be safe. A normal record should never get anywhere close to 10 and if for some reason it does, that's fine.

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py (hunk 2 above)
Comment (mberacochea) [reply helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:186]: like this:
```python
logging.warning(f"cmscan file for accession {mgnify_accession} doesn't exist. Expected file path {deoverlap_path}. Skipping.")
```

Comment (tgurbich) [reply helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py]: Added in https://github.com/EBI-Metagenomics/genomes-catalogue-pipeline/commit/e5acaceef930be1cd356919455fc895693acf87a

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py (hunk 4 above)
Comment (tgurbich) [reply helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py]: Good point. This was meant to check if the accession belongs to a wgs_set but they can now start with other letters too. Modified the code here https://github.com/EBI-Metagenomics/genomes-catalogue-pipeline/commit/0a4d6a5c640f2764005ad7da9a823f494f10e639

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py (hunk 5 above)
Comment (tgurbich) [reply helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:523]: It should exit because this means something is wrong with the ENA request. There might be some edge case where it's not an ENA issue but I haven't encountered that yet and it's better to deal with it properly if it comes up.

File: helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py (hunk 6 above)
Comment (tgurbich) [reply helpers/database-import-scripts/rnacentral/generate_rnacentral_json.py:558]: I changed it to a more manual approach because I couldn't get this to work for my purpose (tracking missing samples).
